### PR TITLE
Prepare C code for compiling on Windows

### PR DESCRIPTION
**Pull request by RobAtticus**
For all exported functions the macro PGDLLEXPORT needs to be pre-
pended. Additionally, on Windows `open` is a macro that needed to
be renamed. A few other small changes are done to make Visual
Studio's compiler happy / get rid of warnings (e.g. adding return
statements after elog.

### CODE REVIEW

File: src/hypertable_cache.c [hunk 1]
```diff
@@ -14,10 +14,10 @@ static void *hypertable_cache_create_entry(Cache *cache, CacheQuery *query);
 
 typedef struct HypertableCacheQuery
 {
-	CacheQuery	q;
 	Oid			relid;
 	const char *schema;
 	const char *table;
+	CacheQuery	q;
```
Comment (RobAtticus) [inline src/hypertable_cache.c]: VS doesn't like something that contains 0-sized arrays to be anything but the last item in a struct.

Comment (erimatnor) [reply src/hypertable_cache.c]: PostgreSQL has a define `FLEXIBLE_ARRAY_MEMBER` to deal with this I think. Look in `pg_config.h.win32` or search the headers to see if this helps.

Comment (RobAtticus) [reply src/hypertable_cache.c]: Thanks, updated.

File: src/cache.h [hunk 2]
```diff
@@ -5,10 +5,17 @@
 #include <utils/memutils.h>
 #include <utils/hsearch.h>
 
+#undef FLEXIBLE_ARRAY_MEMBER
```
Comment (cevian) [inline src/cache.h]: Why do you need to redefine this. Shouldn't pg_config.h set this correctly for the platform.

File: src/version.c [hunk 3]
```diff
@@ -2,8 +2,15 @@
 #include <string.h>
 #include "fmgr.h"
 
+#ifndef EXT_GIT_COMMIT
```
Comment (cevian) [inline src/version.c]: Can we make this actually work? i.e. make the right commit appear in here. We'll probably need this and a #define for the extension version too. Also need to only set "windows" if _WIN32.

File: src/cache.h (hunk 2 above)
Comment (RobAtticus) [reply src/cache.h]: It is not correctly set on the platform (it's not really platform-dependent, it's compiler dependent). On both OSX & Windows at least it is defined as empty.

File: src/version.c (hunk 3 above)
Comment (RobAtticus) [reply src/version.c]: I guess we could but it's really a dev thing used when testing non-release versions. As none of our current developers develop on Windows predominantly, I just put in this hack.

File: src/cache.h (hunk 2 above)
Comment (cevian) [reply src/cache.h]: if you look at ./configure in postgres src it has logic for this. Why are we not hitting it?

Comment (RobAtticus) [reply src/cache.h]: We don't use ./configure on Windows, and when they built the source they may have used a compiler than supports it?

Comment (cevian) [reply src/cache.h]: I would like @olofr and @erimatnor to look at this.

File: src/version.c (hunk 3 above)
Comment (olofr) [reply src/version.c]: I think we should fix this for release builds as well as windows builds but until then I would prefer to set it to `unknown` or something similar.

File: src/cache.h (hunk 2 above)
Comment (olofr) [reply src/cache.h]: I am not sure this is OK to do. Shouldn't this be set somewhere like `﻿﻿﻿src/tools/msvc/config_default.pl` as that seems to take `./configures` role under windows ?

File: src/version.c (hunk 3 above)
Comment (RobAtticus) [reply src/version.c]: Removed and now defined via Visual Studio compiler's command line

File: src/cache.h [hunk 4]
```diff
@@ -5,10 +5,14 @@
 #include <utils/memutils.h>
 #include <utils/hsearch.h>
 
+#ifndef TS_FLEXIBLE_ARRAY_MEMBER
```
Comment (RobAtticus) [inline src/cache.h]: I define our own macro so I can define it in the VS build file correctly regardless of whether the user has PostgreSQL source or just the binary to build against. That is, if they have the binary FLEXIBLE_ARRAY_MEMBER is not necessarily set to 1, which Visual Studio needs. Trying to set FLEXIBLE_ARRAY_MEMBER via command line throws redefinition errors, so this seems like the best solution.

File: Makefile [hunk 5]
```diff
@@ -113,7 +113,7 @@ PGXS := $(shell $(PG_CONFIG) --pgxs)
 EXTRA_CLEAN = $(EXT_SQL_FILE) $(DEPS)
 
 include $(PGXS)
-override CFLAGS += -DINCLUDE_PACKAGE_SUPPORT=0 -MMD -DEXT_GIT_COMMIT=\"$(EXT_GIT_COMMIT)\"
+override CFLAGS += -DINCLUDE_PACKAGE_SUPPORT=0 -MMD -DEXT_GIT_COMMIT=\"$(EXT_GIT_COMMIT)\" -DTS_FLEXIBLE_ARRAY_MEMBER=0
```
Comment (cevian) [inline Makefile]: Why is this necessary? shouldn't #define TS_FLEXIBLE_ARRAY_MEMBER FLEXIBLE_ARRAY_MEMBER work in *unix?

Comment (RobAtticus) [reply Makefile]: clang gave me a warning without it:
```
src/hypertable_cache.c:17:13: warning: field 'q' with variable sized type 'CacheQuery'
      (aka 'struct CacheQuery') not at the end of a struct or class is a GNU extension
      [-Wgnu-variable-sized-type-not-at-end]
        CacheQuery      q;
```

Comment (RobAtticus) [reply Makefile]: It seems like it's defined on my system to be blank rather than 0.

File: src/scanner.c [hunk 6]
```diff
@@ -127,18 +127,18 @@ index_scanner_close(InternalScannerCtx *ctx)
  */
 static Scanner scanners[] = {
 	[ScannerTypeHeap] = {
-		.open = heap_scanner_open,
+		.opentable = heap_scanner_open,
```
Comment (olofr) [inline src/scanner.c]: Sorry, I missed this before. Why add table ? It doesn't make a lot of sense as it is a common interface both for an index scan and a table scan.

Comment (RobAtticus) [reply src/scanner.c]: `open` is a macro on Windows. Happy to change to another name, but `open` is not usable.

Comment (RobAtticus) [reply src/scanner.c]: `openscanner`

Comment (cevian) [reply src/scanner.c]: openscanner works for me.

File: Makefile (hunk 5 above)
Comment (erimatnor) [reply Makefile]: You've actually discovered a bug. The issue is that CacheQuery is not the last member of HypertableCacheQuery, which it needs to be if it has a flexible array member. And HypertableCacheQuery needs to be allocated with space to account for that embeded struct.

So, we should fix the bug first and then you need not define the separate TS_FLEXIBLE_ARRAY_MEMBER and instead we can just use FLEXIBLE_ARRAY_MEMBER

File: src/scanner.c (hunk 6 above)
Comment (erimatnor) [reply src/scanner.c]: openheap or heapopen? (it is actually calling `heap_open()`)

File: Makefile (hunk 5 above)
Comment (RobAtticus) [reply Makefile]: That's what I thought but then I moved it to the end and broke a bunch of the tests and someone suggested FLEXIBLE_ARRAY_MEMBER. But I will try and fix those errors and move it to the end.

Comment (erimatnor) [reply Makefile]: We can probably replace that data array with just an `void *data` pointer and let whoever wants to pass on data allocate that on the heap. In fact, since we now only have the hypertable cache I don't think the data field is even used. But might be useful to keep as a `void *` pointer for future.

Comment (erimatnor) [reply Makefile]: So, CacheQuery actually needs to be first in HypertableCacheQuery since it is casting the "derived" object to the base type (CacheQuery). The solution is to do the above thing with simply getting rid of the flexible data array and just turn it into a void pointer.

Comment (RobAtticus) [reply Makefile]: Yeah that solution is a lot simpler and compiles on both. Thanks!
